src/Chapter3.hs

```diff
@@ -480,6 +519,60 @@ After defining the city, implement the following functions:
    and at least 10 living people inside
 -}
 
+data House 
+  = One 
+  | Two 
+  | Three 
+  | Four
+  deriving (Show, Enum)
+
+data MainBuilding 
+  = Church
+  | Library
+
+data Castle = Castle
+  { castleName :: String
+  } deriving (Show)
+
+data CityCastle
+  = None
+  | OnlyCastle Castle
+  | CastelAndWalls Castle
+
+data MagicalCity = MagicalCity 
+  { cityCastle :: CityCastle
+  , cityMainBuilding :: MainBuilding
+  , cityHouses :: [House]
+  }
+
+buildCastle :: MagicalCity -> CityCastle -> MagicalCity
+buildCastle city castle = case cityCastle city of
+  None -> city { cityCastle = castle }
+  _ -> MagicalCity {cityCastle = castle, cityMainBuilding = cityMainBuilding city, cityHouses = cityHouses city}
```

Comment:
In this case you can also use record-update syntax to set only the `cityCastle` field, as other field didn't change. And you can notice that both cases behave in the same way.
```suggestion
  _ -> city { cityCastle = castle }
```

With your solution it's possible to destroy castle and walls or only walls by passing `OnlyCastle` or `CastleAndWalls`, so this function doesn't always build :smile: It was supposed that the function will take `String` -- castle name -- and build the castle with the given name or change the name of the existing castle. Though, it shouldn't destroy walls :slightly_smiling_face: